File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} n
 * @return {ListNode}
 */
var removeNthFromEnd = function(head, n) {
    let dummy = new ListNode(0, head);
    let first = dummy;
    let second = dummy;

    // Move first pointer n nodes ahead in the linked list
    for (let i = 0; i < n + 1; i++) {
        first = first.next;
    }

    // Move first to the end, maintaining the gap
    while (first !== null) {
        first = first.next;
        second = second.next;
    }

    // Skip the desired node
    second.next = second.next.next;

    // Return the new head node
    return dummy.next;
};

File: 0019-remove-nth-node-from-end-of-list/0019-remove-nth-node-from-end-of-list.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
  dummy = ListNode.new(0, head)
  first = dummy
  second = dummy

  # Move first pointer n nodes ahead in the linked list
  (n + 1).times do
    first = first.next
  end

  # Move first to the end, maintaining the gap
  while first
    first = first.next
    second = second.next
  end

  # Skip the desired node
  second.next = second.next.next

  # Return the new head node
  dummy.next
end